posix/unistd.c

```diff
@@ -0,0 +1,606 @@
+    oe_pid_t ret = 0;
+    oe_posix_getppid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getpgrp(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getpgrp(&ret);
+    return ret;
+}
+
+oe_uid_t oe_getuid(void)
+{
+    oe_uid_t ret = 0;
+    oe_posix_getuid(&ret);
+    return ret;
+}
+
+oe_uid_t oe_geteuid(void)
+{
+    oe_uid_t ret = 0;
+    oe_posix_geteuid(&ret);
+    return ret;
+}
+
+oe_gid_t oe_getgid(void)
+{
+    oe_gid_t ret = 0;
+    oe_posix_getgid(&ret);
+    return ret;
+}
+
+oe_gid_t oe_getegid(void)
+{
+    oe_gid_t ret = 0;
+    oe_posix_getegid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getpgid(oe_pid_t pid)
+{
+    oe_pid_t ret = -1;
+    oe_pid_t retval = -1;
+
+    if (oe_posix_getpgid(&retval, pid) != OE_OK)
+    {
+        oe_errno = OE_EINVAL;
+        goto done;
+    }
+
+    if (retval == -1)
+        goto done;
+
+    ret = retval;
+
+done:
+    return ret;
+}
+
+int oe_getgroups(int size, oe_gid_t list[])
+{
+    int ret = -1;
+    int retval = -1;
+
+    if (oe_posix_getgroups(&retval, (size_t)size, list) != OE_OK)
+    {
+        oe_errno = OE_EINVAL;
+        goto done;
+    }
+
+    if (retval == -1)
+    {
+        goto done;
+    }
+
+    ret = retval;
+
+done:
+    return ret;
+}
+
+ssize_t oe_read(int fd, void* buf, size_t count)
+{
+    ssize_t ret = -1;
+    oe_device_t* device;
+
+    if (!(device = oe_fdtable_get(fd, OE_DEVICE_TYPE_ANY)))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    ret = OE_CALL_BASE(read, device, buf, count);
+
+done:
+    return ret;
+}
+
+ssize_t oe_write(int fd, const void* buf, size_t count)
+{
+    ssize_t ret = -1;
+    oe_device_t* device;
+
+    if (!(device = oe_fdtable_get(fd, OE_DEVICE_TYPE_ANY)))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    ret = OE_CALL_BASE(write, device, buf, count);
+
+done:
+    return ret;
+}
+
+int oe_close(int fd)
+{
+    int ret = -1;
+    oe_device_t* device;
+
+    if (!(device = oe_fdtable_get(fd, OE_DEVICE_TYPE_ANY)))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    if ((ret = OE_CALL_BASE(close, device)) == 0)
+        oe_fdtable_release(fd);
+
+done:
+    return ret;
+}
+
+int oe_dup(int oldfd)
+{
+    int ret = -1;
+    oe_device_t* old_dev;
+    oe_device_t* new_dev = NULL;
+    int newfd;
+
+    if (!(old_dev = oe_fdtable_get(oldfd, OE_DEVICE_TYPE_ANY)))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    if (OE_CALL_BASE(dup, old_dev, &new_dev) == -1)
+        OE_RAISE_ERRNO(oe_errno);
+
+    if ((newfd = oe_fdtable_assign(new_dev)) == -1)
+        OE_RAISE_ERRNO(oe_errno);
+
+    ret = newfd;
+    new_dev = NULL;
+
+done:
+
+    if (new_dev)
+        OE_CALL_BASE(close, new_dev);
+
+    return ret;
+}
+
+int oe_dup2(int oldfd, int newfd)
+{
+    oe_device_t* old_dev;
+    oe_device_t* new_dev = NULL;
+    int retval = -1;
+
+    if (!(old_dev = oe_fdtable_get(oldfd, OE_DEVICE_TYPE_ANY)))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    if ((retval = OE_CALL_BASE(dup, old_dev, &new_dev)) < 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    if (oe_fdtable_reassign(newfd, new_dev) == -1)
```

Comment:
[POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/functions/dup2.html) seems to require that `dup2()` return `newfd` without closing it if it is equal to `oldfd`. It's not clear to me that this produces the expected result (it reads like it should not dupe in that case).

Reply:
Thanks Simon for finding that. I checked this and if **oldfd** is equal to **newfd**, then the function should simply return **newfd** (or **oldfd**). Fixed.